### PR TITLE
Fix issue 6904 by removing the subclass parameter from model auto complete

### LINKED ISSUE #6904: An exception has been thrown during the rendering of a template ("Feature not implemented: an embedded admin cannot have subclass").
#### Sonata packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle              3.87.0 3.91.0 The missing Symfony Admin Generator
sonata-project/block-bundle              3.21.0 4.5.0  Symfony SonataBlockBundle
sonata-project/cache                     2.0.1  2.1.1  Cache library
sonata-project/doctrine-extensions       1.9.1  1.11.0 Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.29.0 3.30.0 Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/exporter                  2.5.1  2.5.1  Lightweight Exporter library
sonata-project/form-extensions           1.6.0  1.9.0  Symfony form extensions
sonata-project/media-bundle              3.27.0 3.30.0 Symfony SonataMediaBundle
sonata-project/translation-bundle        2.7.0  2.7.0  SonataTranslationBundle
sonata-project/twig-extensions           1.5.1  1.5.1  Sonata twig extensions

```

</p>
</details>

#### Symfony packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'symfony/*'
symfony/asset                      v4.4.19 v5.2.3  Manages URL generation and versioning of web assets such as CSS stylesheets, JavaS...
symfony/browser-kit                v4.4.19 v5.2.3  Simulates the behavior of a web browser, allowing you to make requests, click on l...
symfony/cache                      v5.2.3  v5.2.3  Provides an extended PSR-6, PSR-16 (and tags) implementation
symfony/cache-contracts            v2.2.0  v2.2.0  Generic abstractions related to caching
symfony/config                     v4.4.19 v5.2.3  Helps you find, load, combine, autofill and validate configuration values of any kind
symfony/console                    v4.4.19 v5.2.3  Eases the creation of beautiful and testable command line interfaces
symfony/css-selector               v4.4.19 v5.2.3  Converts CSS selectors to XPath expressions
symfony/debug                      v4.4.19 v4.4.19 Provides tools to ease debugging PHP code
symfony/debug-bundle               v4.4.19 v5.2.3  Provides a tight integration of the Symfony Debug component into the Symfony full-...
symfony/dependency-injection       v4.4.19 v5.2.3  Allows you to standardize and centralize the way objects are constructed in your a...
symfony/deprecation-contracts      v2.2.0  v2.2.0  A generic function and convention to trigger deprecation notices
symfony/doctrine-bridge            v4.4.19 v5.2.3  Provides integration for Doctrine with various Symfony components
symfony/dom-crawler                v4.4.19 v5.2.3  Eases DOM navigation for HTML and XML documents
symfony/dotenv                     v4.4.19 v5.2.3  Registers environment variables from a .env file
symfony/error-handler              v4.4.19 v5.2.3  Provides tools to manage errors and ease debugging PHP code
symfony/event-dispatcher           v4.4.19 v5.2.3  Provides tools that allow your application components to communicate with each oth...
symfony/event-dispatcher-contracts v1.1.9  v2.2.0  Generic abstractions related to dispatching event
symfony/expression-language        v5.2.3  v5.2.3  Provides an engine that can compile and evaluate expressions
symfony/filesystem                 v5.2.3  v5.2.3  Provides basic utilities for the filesystem
symfony/finder                     v4.4.19 v5.2.3  Finds files and directories via an intuitive fluent interface
symfony/flex                       v1.12.2 v1.12.2 Composer plugin for Symfony
symfony/form                       v4.4.19 v5.2.3  Allows to easily create, process and reuse HTML forms
symfony/framework-bundle           v4.4.19 v5.2.3  Provides a tight integration between Symfony components and the Symfony full-stack...
symfony/http-client                v4.4.13 v5.2.3  Symfony HttpClient component
symfony/http-client-contracts      v2.3.1  v2.3.1  Generic abstractions related to HTTP clients
symfony/http-foundation            v4.4.19 v5.2.3  Defines an object-oriented layer for the HTTP specification
symfony/http-kernel                v4.4.19 v5.2.3  Provides a structured process for converting a Request into a Response
symfony/inflector                  v5.2.3  v5.2.3  Converts words between their singular and plural forms (English only)
symfony/intl                       v4.4.19 v5.2.3  Provides a PHP replacement layer for the C intl extension that includes additional...
symfony/mailer                     v4.4.13 v5.2.3  Symfony Mailer Component
symfony/maker-bundle               v1.29.1 v1.29.1 Symfony Maker helps you create empty commands, controllers, form classes, tests an...
symfony/messenger                  v4.4.19 v5.2.3  Helps applications send and receive messages to/from other applications or via mes...
symfony/mime                       v5.2.3  v5.2.3  Allows manipulating MIME messages
symfony/monolog-bridge             v4.4.19 v5.2.3  Provides integration for Monolog with various Symfony components
symfony/monolog-bundle             v3.6.0  v3.6.0  Symfony MonologBundle
symfony/options-resolver           v4.4.19 v5.2.3  Provides an improved replacement for the array_replace PHP function
symfony/orm-pack                   v1.0.8  v2.1.0  A pack for the Doctrine ORM
symfony/phpunit-bridge             v5.2.3  v5.2.3  Provides utilities for PHPUnit, especially user deprecation notices management
symfony/polyfill-ctype             v1.22.1 v1.22.1 Symfony polyfill for ctype functions
symfony/polyfill-intl-grapheme     v1.22.1 v1.22.1 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu          v1.22.1 v1.22.1 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn          v1.22.1 v1.22.1 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer   v1.22.1 v1.22.1 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.22.1 v1.22.1 Symfony polyfill for the Mbstring extension
symfony/polyfill-php72             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/polyfill-php80             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/process                    v4.4.19 v5.2.3  Executes commands in sub-processes
symfony/profiler-pack              v1.0.5  v1.0.5  A pack for the Symfony web profiler
symfony/property-access            v4.4.19 v5.2.3  Provides functions to read and write from/to an object or array using a simple str...
symfony/routing                    v4.4.19 v5.2.3  Maps an HTTP request to a set of configuration variables
symfony/security                   v4.4.19 v4.4.19 Provides a complete security system for your web application
symfony/security-acl               v3.1.1  v3.1.1  Symfony Security Component - ACL (Access Control List)
symfony/security-bundle            v4.4.19 v5.2.3  Provides a tight integration of the Security component into the Symfony full-stack...
symfony/serializer                 v4.4.19 v5.2.3  Handles serializing and deserializing data structures, including object graphs, in...
symfony/service-contracts          v2.2.0  v2.2.0  Generic abstractions related to writing services
symfony/stopwatch                  v5.2.3  v5.2.3  Provides a way to profile code
symfony/string                     v5.2.3  v5.2.3  Provides an object-oriented API to strings and deals with bytes, UTF-8 code points...
symfony/swiftmailer-bundle         v3.5.2  v3.5.2  Symfony SwiftmailerBundle
symfony/templating                 v4.4.19 v5.2.3  Provides all the tools needed to build any kind of template system
symfony/test-pack                  v1.0.7  v1.0.7  A pack for functional and end-to-end testing within a Symfony app
symfony/translation                v4.4.19 v5.2.3  Provides tools to internationalize your application
symfony/translation-contracts      v2.3.0  v2.3.0  Generic abstractions related to translation
symfony/twig-bridge                v4.4.19 v5.2.3  Provides integration for Twig with various Symfony components
symfony/twig-bundle                v4.4.19 v5.2.3  Provides a tight integration of Twig into the Symfony full-stack framework
symfony/validator                  v4.4.19 v5.2.3  Provides tools to validate values
symfony/var-dumper                 v4.4.19 v5.2.3  Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter               v5.2.3  v5.2.3  Allows exporting any serializable PHP data structure to plain PHP code
symfony/web-profiler-bundle        v5.0.11 v5.2.3  Symfony WebProfilerBundle
symfony/workflow                   v5.2.3  v5.2.3  Provides tools for managing a workflow or finite state machine
symfony/yaml                       v4.4.19 v5.2.3  Loads and dumps YAML files

```

</p>
</details>

#### PHP version

```
$ php -v
PHP 7.4.2 (cli) (built: Feb  1 2020 19:47:36) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.2, Copyright (c), by Zend Technologies
    with Xdebug v2.9.8, Copyright (c) 2002-2020, by Derick Rethans

```

## Steps to reproduce

I have an admin with inheritance and embedded admin. The embedded admin don't have sub class.

I think the check is now done on the main admin and not the embedded admin.

**This bug was introduce int version 3.87.0 (I have check and it's working on version 3.86.0 and still don't work in latest tag)**

## Expected results

No error

## Actual results

I am getting this error:

An exception has been thrown during the rendering of a template ("Feature not implemented: an embedded admin cannot have subclass").

### PR DESCRIPTION
**Pull request by mpoiriert**
…

## Subject

I am targeting this branch, because the bug was introduce here.

Closes #6904.

## Changelog

```markdown
### Fixed
Model Auto Complete failing in create flow of admin with subclasses
```

### CODE REVIEW

**Comment by mpoiriert** [comment]
@VincentLanglet Ready to review. The automation test are just checking that a subclass value is not there,. I don't think I can do more than this.

**Comment by VincentLanglet** [comment]
> The automation test are just checking that a subclass value is not there,. I don't think I can do more than this.

Best would be a functional test to reproduce the issue from #6904. Do you think it would be possible ?

**Comment by mpoiriert** [comment]
The issue is base on the fact that the subclass parameter is present when
we are on creation flow with an admin that have subclass calling the model
auto complete endpoint via ajax. Not sure how you would proceed to do that
without a panther setup ?

Just calling the model auto complete endpoint without having a subclass
parameter don't test anything since I didn't change anything other than
removing the parameter in the twig template.

The fix I made for the other issue was to remove/put back the parameter in
the controller but that was a "hack" since we had too many place to check.
The best solution would have been to remove the parameter completely which
the only wau to test would have been to check that the parameter is not
there.

**Comment by VincentLanglet** [comment]
> The fix I made for the other issue was to remove/put back the parameter in the controller but that was a "hack" since we had too many place to check. The best solution would have been to remove the parameter completely which the only wau to test would have been to check that the parameter is not there.

Does the other fix still needed then ?

**Comment by mpoiriert** [comment]
Yes since it's not the same endpoint.

Once we found all the subclass parameter that have issues we could revisit how the fix was done.

The current one seems to be a "final" fix for that specific use case.

**Review by VincentLanglet — DISMISSED** [review]

File: tests/Functional/Controller/CRUDControllerTest.php [hunk 1]
```diff
@@ -34,8 +34,9 @@ public function testList(): void
 
     public function testCreate(): void
     {
+        $subclass = uniqid('subclass');
         $client = static::createClient();
-        $crawler = $client->request(Request::METHOD_GET, '/admin/tests/app/foo/create');
+        $crawler = $client->request(Request::METHOD_GET, '/admin/tests/app/foo/create?subclass='.$subclass);
```
Comment (franmomu) [inline tests/Functional/Controller/CRUDControllerTest.php]: IMHO this should go in a new test instead of modifying this one

Comment (VincentLanglet) [reply tests/Functional/Controller/CRUDControllerTest.php]: Indeed @mpoiriert, two tests are better than one

**Review by VincentLanglet — DISMISSED** [review]

File: tests/Functional/Controller/CRUDControllerTest.php [hunk 2]
```diff
@@ -48,6 +48,24 @@ public function testCreate(): void
         );
     }
 
+    /**
+     * https://github.com/sonata-project/SonataAdminBundle/issues/6904
```
Comment (VincentLanglet) [inline tests/Functional/Controller/CRUDControllerTest.php]:
```suggestion
     * https://github.com/sonata-project/SonataAdminBundle/issues/6904.
```
For the linter

Comment (mpoiriert) [reply tests/Functional/Controller/CRUDControllerTest.php]: Done

File: tests/Functional/Controller/CRUDControllerTest.php (hunk 1 above)
Comment (mpoiriert) [reply tests/Functional/Controller/CRUDControllerTest.php]: Done

**Review by VincentLanglet — APPROVED** [review]

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Thanks @mpoiriert

**Comment by Geekimo** [comment]
Hello,
following this PR, we have issues when editing subjects that are "abstract" and use a discriminator map.
I've found a workaround by passing the subclass parameter as `req_params` in `ModelAutocompleteType::class`.
Shall we found another way to fix @mpoiriert issue that is less "global" ?

**Comment by mpoiriert** [comment]
@Geekimo The issue is the way the subclass have been implemented.
What is your specific use case ? Because I removed the sublcass use from the query parameter which is conflicting. If you do an Edit the query parameter subclass is not set since it's just when you do a create.

**Comment by Geekimo** [comment]
@mpoiriert Our admin class can edit two entities inheriting from the same abstract entity. We found out by going back to the latest version before your fix that the `subclass` query parameter was missing in the ajax request, and I had to set it by hand.

**Comment by mpoiriert** [comment]
@Geekimo To make sure I understand you were already making a "hook" since the behaviour wasn't suppose to accomplish what you were trying to do ?

**Comment by Geekimo** [comment]
@mpoiriert  AFAIK there wasn't any hook.
Maybe it has never worked for edit and nobody noticed it . Strange.
Thanks for the reply.

**Comment by jordisala1991** [comment]
Not sure if I follow correctly the conversation. Something broke after this PR?

**Comment by mpoiriert** [comment]
It's not the same endpoint. So yes. Once we found all the subclass
parameter that have issues we could revisit how the fix was done. I'll
check this later and let you know.

**Comment by ENDaZONELT** [comment]
@mpoiriert any news regarding this bug ? I have the same issue. Abstract entity and discriminator map. I need to pass subclass parameter manually via req_params:
'req_params' => ['subclass' => array_search(get_class($this->getSubject()), $this->getSubClasses())],

**Comment by mpoiriert** [comment]
@ENDaZONELT The issue was that subclass should not be passed automatically. I don't know your exact use case but the one that I fix was because AdminA (wiht subclass) -> Calling AutoComplete of AdminB was not working because the subclass of admin A was pass to AdminB.

**Pull request merged**